htsget-search/src/storage/local.rs

```diff
@@ -1,34 +1,96 @@
 //! Module providing an implementation for the [Storage] trait using the local file system.
 //!
 
+use std::path::{Path, PathBuf};
+
 use async_trait::async_trait;
 use tokio::fs::File;
 
+use htsget_id_resolver::{HtsGetIdResolver, RegexResolver};
+
 use crate::htsget::Url;
-use crate::storage;
 use crate::storage::async_storage::AsyncStorage;
-use crate::storage::blocking::local::LocalStorage;
 
 use super::{GetOptions, Result, StorageError, UrlOptions};
 
-#[async_trait]
-impl AsyncStorage for LocalStorage {
-  type Streamable = File;
+/// Implementation for the [Storage] trait using the local file system.
+#[derive(Debug)]
+pub struct LocalStorage {
+  base_path: PathBuf,
+  id_resolver: RegexResolver,
+}
 
-  async fn get<K: AsRef<str> + Send>(&self, key: K, _options: GetOptions) -> Result<File> {
+impl LocalStorage {
+  pub fn new<P: AsRef<Path>>(base_path: P, id_resolver: RegexResolver) -> Result<Self> {
+    base_path
+      .as_ref()
+      .to_path_buf()
+      .canonicalize()
+      .map_err(|_| StorageError::KeyNotFound(base_path.as_ref().to_string_lossy().to_string()))
+      .map(|canonicalized_base_path| Self {
+        base_path: canonicalized_base_path,
+        id_resolver,
+      })
+  }
+
+  pub fn base_path(&self) -> &Path {
+    self.base_path.as_path()
+  }
+
+  pub(crate) fn get_path_from_key<K: AsRef<str>>(&self, key: K) -> Result<PathBuf> {
+    let key: &str = key.as_ref();
+    self
+      .base_path
+      .join(
+        self
+          .id_resolver
+          .resolve_id(key)
+          .ok_or_else(|| StorageError::InvalidKey(key.to_string()))?,
+      )
+      .canonicalize()
+      .map_err(|_| StorageError::InvalidKey(key.to_string()))
+      .and_then(|path| {
+        path
+          .starts_with(&self.base_path)
+          .then(|| path)
+          .ok_or_else(|| StorageError::InvalidKey(key.to_string()))
+      })
+      .and_then(|path| {
+        path
+          .is_file()
+          .then(|| path)
+          .ok_or_else(|| StorageError::KeyNotFound(key.to_string()))
+      })
+  }
+
+  async fn get<K: AsRef<str>>(&self, key: K) -> Result<File> {
     let path = self.get_path_from_key(&key)?;
     File::open(path)
       .await
       .map_err(|e| StorageError::IoError(e, key.as_ref().to_string()))
   }
+}
+
+#[async_trait]
+impl AsyncStorage for LocalStorage {
+  type Streamable = File;
+
+  /// Get the file at the location of the key.
+  async fn get<K: AsRef<str> + Send>(&self, key: K, _options: GetOptions) -> Result<File> {
+    self.get(key).await
+  }
 
+  /// Get a url for the file at key.
   async fn url<K: AsRef<str> + Send>(&self, key: K, options: UrlOptions) -> Result<Url> {
-    storage::blocking::Storage::url(self, key, options)
+    // TODO file:// is not allowed by the spec. We should consider including an static http server for the base_path
```

Comment:
I think it's fundamental to address this now before it gets more hairy. As @CastilloDel pointed out in https://github.com/umccr/htsget-rs/issues/73, this also affects the benchmarking code/PRs... so I think it'd make sense to solve it, if not in this PR, right afterwards.

Since we are already using tokio as a Noodles dependency, would it make sense to have Axum as that static/testing http server to be able to fully comply with the htsget spec?

Reply:
Yeah, I think any server should be okay. 
Do you think it makes sense to start another server instance in the Storage interface, separate from the main server that handles the queries?

Reply:
My intuition tells me that it'd be better to have just one (http) server handling it all, but I might be missing something you spotted?

Reply:
I might be a little bit easier to implement as two separate servers and make it more modular. That way the query server (i.e. htsget-http-actix) doesn't have to worry about local paths, and storage will always return valid urls. It also means that if there is another server crate that isn't actix, it doesn't have to re-implement serving the static local files.

Reply:
Gotcha & go for it!